File: day13/main.dart
import 'dart:async';
import 'dart:io';
import 'dart:convert';

Future<List<String>> getLines(filename) async {
  List<String> lines = [];
  await new File(filename)
      .openRead()
      .map(utf8.decode)
      .transform(new LineSplitter())
      .forEach((line) {
    lines.add(line);
  });
  return lines;
}

void main(List<String> arguments) async {
  String filename = 'input.txt';
  if (arguments.length == 1) {
    filename = arguments[0];
  }

  int part1 = 0;
  int part2 = 0;

  List<String> lines = await getLines(filename);
  List<List<int>> coords = [];
  List<String> folds = [];

  for (final line in lines) {
    if (line.isNotEmpty) {
      if (line.length >= 13) {
        folds.add(line.split(' ')[2]);
      } else {
        coords.add(line.split(',').map(int.parse).toList());
      }
    }
  }

  int maxWidth = 0, maxHeight = 0;
  for (final coord in coords) {
    if (coord[0] > maxWidth) {
      maxWidth = coord[0];
    }
    if (coord[1] > maxHeight) {
      maxHeight = coord[1];
    }
  }
  List<List<String>> manual = List.generate(
      maxHeight + 1, (i) => List.generate(maxWidth + 1, (j) => '.'));

  int count = 0;
  for (final coord in coords) {
    manual[coord[1]][coord[0]] = count.toString();
    count++;
  }

  manual.forEach((line) {
    print(line);
  });

  String fold = folds[0];
  if (fold[0] == 'x') {
    int foldindex = int.parse(fold.split('=')[1]);
    for (final line in manual) {
      for (int index = foldindex; index < line.length; index++) {
        if (line[index] != '.') {
          print('$index, ${index - foldindex}');
          line[index - foldindex] = line[index];
        }
      }
      line.length = foldindex;
    }
  }

  print('\n\n');
  manual.forEach((line) {
    line.forEach((element) {
      if (element == '#') {
        part1++;
      }
    });
    print('$line');
  });

  print('part 1: $part1');
  print('part 2: $part2');
}

File: day13/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type dot struct {
	x int
	y int
}

func main() {
	var filename string = "input.txt"
	if len(os.Args) == 2 {
		filename = os.Args[1]
	}
	input, err := os.ReadFile(filename)
	if err != nil {
		panic(err)
	}

	var dots = []dot{}
	lines := strings.Trim(string(input), "\n")
	for _, coords := range strings.Split(lines, "\n") {
		if strings.Contains(coords, ",") {
			c := strings.Split(coords, ",")
			fmt.Println(c[0], c[1])
			d := dot{val, _ = strconv.ParseInt(c[0], 10, 16); val, strconv.ParseInt(c[1], 10, 16)}
			append(dots, d)

		} else if strings.Contains(coords, "x") {
			fmt.Println(strings.Split(coords, "=")[1])
		}
	}
	fmt.Println(dots)
}
